src/app/api/collections/counts.py

```diff
@@ -0,0 +1,78 @@
+from enum import Enum
+from uuid import UUID
+
+from elasticsearch_dsl import A
+from pydantic import BaseModel
+
+from app.core.config import ELASTIC_TOTAL_SIZE
+from app.elastic.elastic import query_materials
+from app.elastic.search import Search
+from app.models import CollectionAttribute, ElasticResourceAttribute
+
+
+class PortalTreeCount(BaseModel):
+    noderef_id: UUID
+    total: int
+    counts: dict[str, int]
+
+
+_AGGREGATION_NAME = "collection_id"
+
+
+class AggregationMappings(str, Enum):
+    """
+    Mappings of the elastic fields where facets should be built for each individual collection
+    """
+
+    lrt = ("properties.ccm:oeh_lrt_aggregated.keyword",)
+    license = ("properties.ccm:commonlicense_key.keyword",)
+
+
+def query_portal_counts(node_id: UUID, facet: AggregationMappings) -> Search:
+    s = Search().base_filters().query(query_materials(ancestor_id=node_id))
+    material_agg = A(
+        "terms", field="collections.nodeRef.id.keyword", size=ELASTIC_TOTAL_SIZE
+    )
+    material_agg.bucket(
+        "facet",
+        A(
+            "terms",
+            field=facet,
+            size=ELASTIC_TOTAL_SIZE,
+        ),
+    )
+    s.aggs.bucket(_AGGREGATION_NAME, material_agg)
+    s = s.source(
+        [
+            ElasticResourceAttribute.NODEREF_ID.path,
+            CollectionAttribute.TITLE.path,
+            CollectionAttribute.PATH.path,
+            CollectionAttribute.PARENT_ID.path,
+        ]
+    )[:0]
+    return s
+
+
+async def portal_counts(
+    node_id: UUID, facet: AggregationMappings
+) -> list[PortalTreeCount]:
+    response = query_portal_counts(node_id, facet).execute()
+    if response.success():
```

Comment:
if the query fails, then this function does not return a list...
I would rather do an `if not success:  raise ...` and either let the caller deal with the exception, or let the exception (50x gateway error?) propagate upwards towards the fastapi endpoint

Reply:
Interestingly, if nothing is returned here, FastAPI will catch it through Pydantic and yield an exception. If you want, I can raise here explicitly, but we do not have to do this.

Reply:
The fastapi handling for it is good, but still, the type annotation should at least state that the return value may be none, i.e. Optional[list[PortalTreeCount]]

Reply:
Well, now you know my private github account :-)

Reply:
:D Nice name ^^ 

I added the optional, good suggestion!